metrics-scripts/encapsulate_testnet_validator_metrics.py

```diff
@@ -0,0 +1,138 @@
+import os
+import re
+import glob
+import gzip
+import sys
+from datetime import datetime, timedelta
+from dateutil import parser
+import requests
+
+# Configuration
+parent_dir = sys.argv[1]
+log_dir = os.path.join(parent_dir, "log")
+log_pattern = "supra.log"
+
+def find_latest_log_file():
+    """Finds the latest log file based on modification time."""
+    files = sorted(glob.glob(f"{log_dir}/{log_pattern}"), key=os.path.getmtime, reverse=True)
+    return files[0] if files else None
+
+def parse_timestamp_ns(line):
+    """Parses timestamp from log line and converts to nanoseconds."""
+    match = re.search(r"\[([^\]]+)\]", line)
+    if match:
+        timestamp_str = match.group(1).replace("Z+00:00", "")
+        try:
+            timestamp_dt = parser.isoparse(timestamp_str)
+            return int(timestamp_dt.timestamp() * 1e9)
+        except ValueError:
+            return None
+    return None
+
+def calculate_block_rates():
+    """Calculate block round and height time rates."""
+    latest_log = find_latest_log_file()
+    if not latest_log:
+        return None, None
+
+    last_round, last_round_ts = None, None
+    second_last_round, second_last_round_ts = None, None
+    last_height, last_height_ts = None, None
+    second_last_height, second_last_height_ts = None, None
+
+    with open(latest_log, "r") as f:
+        for line in f:
+            ts = parse_timestamp_ns(line)
+            if not ts:
+                continue
+
+            if "Block round" in line:
+                match = re.search(r"Block round: \((\d+)\)", line)
+                if match:
+                    round_num = int(match.group(1))
+                    if last_round is not None:
+                        second_last_round, second_last_round_ts = last_round, last_round_ts
+                    last_round, last_round_ts = round_num, ts
+
+            if "Block height" in line:
+                match = re.search(r"Block height: \((\d+)\)", line)
+                if match:
+                    height = int(match.group(1))
+                    if last_height is not None:
+                        second_last_height, second_last_height_ts = last_height, last_height_ts
+                    last_height, last_height_ts = height, ts
+
+    round_rate = None
+    height_rate = None
+
+    if second_last_round is not None and last_round != second_last_round:
+        round_rate = (last_round_ts - second_last_round_ts) / 1e9
```

Comment:
If you wanna calculate block rate, take a large window and then calculate how many blocks were there in that time duration of the window